src/webgpu/api/validation/vertex_state.spec.ts

```diff
@@ -303,73 +310,169 @@ g.test('vertex_attribute_shaderLocation_limit')
     t.testVertexState(success, { vertexBuffers });
   });
 
-g.test('pipeline_vertex_buffers_are_backed_by_attributes_in_vertex_input').fn(async t => {
-  const vertexState: GPUVertexStateDescriptor = {
-    vertexBuffers: [
+g.test('vertex_attribute_shaderLocation_unique')
+  .desc(
+    `Test that shaderLocation must be unique in the vertex state.
+   - Test for various pairs of buffers that contain the potentially conflicting attributes
+   - Test for the potentially conflicting attributes in various places in the buffers (with dummy attributes)
+   - Test for various shaderLocations that conflict or not`
+  )
+  .subcases(() =>
+    params()
+      .combine(poptions('vertexBufferIndexA', [0, 1, kMaxVertexBuffers - 1]))
+      .combine(poptions('vertexBufferIndexB', [0, 1, kMaxVertexBuffers - 1]))
+      .combine(pbool('testAttributeAtStartA'))
+      .combine(pbool('testAttributeAtStartB'))
+      .combine(poptions('shaderLocationA', [0, 1, 7, kMaxVertexAttributes - 1]))
+      .combine(poptions('shaderLocationB', [0, 1, 7, kMaxVertexAttributes - 1]))
+      .combine(poptions('extraAttributes', [0, 4]))
+  )
+  .fn(t => {
+    const {
+      vertexBufferIndexA,
+      vertexBufferIndexB,
+      testAttributeAtStartA,
+      testAttributeAtStartB,
+      shaderLocationA,
+      shaderLocationB,
+      extraAttributes,
+    } = t.params;
+
+    const vertexBufferAttributes: GPUVertexAttributeDescriptor[][] = [];
+    vertexBufferAttributes[vertexBufferIndexA] = [];
+    vertexBufferAttributes[vertexBufferIndexB] = [];
+
+    // Add the dummy attributes for attribute A
+    const attributesA = vertexBufferAttributes[vertexBufferIndexA];
+    let currentLocation = 0;
+    for (let i = 0; i < extraAttributes; i++) {
+      if (currentLocation === shaderLocationA || currentLocation === shaderLocationB) {
+        currentLocation++;
+        continue;
+      }
+
+      attributesA.push({ format: 'float', shaderLocation: currentLocation, offset: 0 });
+      currentLocation++;
+    }
+
+    // Add attribute A
+    if (testAttributeAtStartA) {
+      attributesA.unshift({ format: 'float', offset: 0, shaderLocation: shaderLocationA });
+    } else {
+      attributesA.push({ format: 'float', offset: 0, shaderLocation: shaderLocationA });
+    }
+
+    // Add attribute B potentially in the same list of attributes as A
+    const attributesB = vertexBufferAttributes[vertexBufferIndexB];
+    if (testAttributeAtStartB) {
```

Comment:
isn't this array empty at this point so this has no effect?

Reply:
It can be the same object as `attrributesA`. Made the comment more clear.